include/clang/AST/Expr.h

```diff
@@ -4514,6 +4514,120 @@ class GenericSelectionExpr : public Expr {
   friend class ASTStmtReader;
 };
 
+/// \brief Represents a Checked C bounds expression.
+class BoundsExpr : public Expr {
```

Comment:
There's lots of code that operates over Expr that is also useful for BoundsExpr.  We'd lose the re-use that code if we made BoundsExpr a peer of Expr in the class hierarchy.